modules/cfe_testcase/src/time_conversion_test.c

```diff
@@ -0,0 +1,172 @@
+ */
+#include "cfe_test.h"
+
+/* Check if its nearly possible that Time == Target */
+/* DUPLICATE! from time_current_test.c */
+bool TimeInRange(CFE_TIME_SysTime_t Time, CFE_TIME_SysTime_t Target, OS_time_t difference)
+{
+	OS_time_t Max;
+	OS_time_t TimeT = OS_TimeAssembleFromSubseconds(Time.Seconds, Time.Subseconds);
+	OS_time_t TargetT = OS_TimeAssembleFromSubseconds(Target.Seconds, Target.Subseconds);
+
+	Max = OS_TimeAdd(TimeT, difference);
+
+	if (TargetT.ticks >= TimeT.ticks && TargetT.ticks <= Max.ticks)
+	{
+		return true;
+	}
+	else
+	{
+		return false;
+	}
+}
+
+void TestConvertMET2SCTime(void)
+{
+	UtPrintf("Testing: CFE_TIME_MET2SCTime");
+
+	/* std os time variables */
+	OS_time_t          start;
+	OS_time_t          end;
+	OS_time_t          difference;
+
+	/* Mission Elapsed Time */
+	CFE_TIME_SysTime_t MET;
+	/* MET + SCTF */
+	CFE_TIME_SysTime_t TAI;
+	/* MET + SCTF - Leap Seconds */
+	CFE_TIME_SysTime_t UTC;
+	/* Spacecraft Time */
+	CFE_TIME_SysTime_t SCTime;
+
+	/* Print buffers */
+	char timeBuf1[sizeof("yyyy-ddd-hh:mm:ss.xxxxx_")];
+	char timeBuf2[sizeof("yyyy-ddd-hh:mm:ss.xxxxx_")];
+
+	OS_GetLocalTime(&start);
+
+	/* Get Times */
+	MET = CFE_TIME_GetMET();
+	TAI = CFE_TIME_GetTAI();
+	UTC = CFE_TIME_GetUTC();
+
+	OS_GetLocalTime(&end);
+
+	/* calculate diffrence */
+	difference = OS_TimeSubtract(end, start);
+
+	/* 
+	 * Convert - should produce a TAI or UTC at the moment of GetMET() 
+	 *			 with a minimal time shift between 
+	 */
+	SCTime = CFE_TIME_MET2SCTime(MET);
+
+	/* write Spacecraft Time into second buffer */
+	CFE_TIME_Print(timeBuf2, SCTime);
+
+	/* Check conversion */
+#if (CFE_MISSION_TIME_CFG_DEFAULT_TAI == true)
+	/* SCTime is TAI format */
+
+	/* avoid unused compiler warning */
+	(void)UTC;
+
+	/* write TAI into first buffer */
+	CFE_TIME_Print(timeBuf1, TAI);
+
+	/*
+	 * CFE_TIME_Compare not possible because of the 
+	 * temporal diffrence between get GetMET and GetTAI 
+	 */
+	UtAssert_True(TimeInRange(TAI, SCTime, difference), "TAI (%s) = MET2SCTime (%s)", timeBuf1, timeBuf2);
+
+	/* this would be ok but it would repeat a whats done in CFE_TIME_MET2SCTime */
+	/* UtAssert_True(CFE_TIME_Compare(SCTime, CFE_TIME_Add(MET, STCF))); */
+#else
+	/* SCTime is UTC format */
+
+	/* avoid unused compiler warning */
+	(void)TAI;
+
+	/* write UTC into first buffer */
+	CFE_TIME_Print(timeBuf1, UTC);
+
+	/* 
+	 * CFE_TIME_Compare not possible because of the
+	 * temporal diffrence between get GetMETand GetUTC 
+	 */
+	UtAssert_True(TimeInRange(UTC, SCTime, difference), "UTC (%s) = MET2SCTime (%s)", timeBuf1, timeBuf2);
+#endif
+}
+
+void TestConvertSubSeconds2MicroSeconds(void)
+{
+	UtPrintf("Testing: CFE_TIME_Sub2MicroSecs");
+
+	uint32 SUB;
+	uint32 MS;
+	OS_time_t TimeSUB;
+
+	SUB = CFE_TIME_GetMETsubsecs();
+	MS = CFE_TIME_Sub2MicroSecs(SUB);
+
+	TimeSUB = OS_TimeAssembleFromSubseconds(0, SUB);
+
+	UtAssert_True(OS_TimeGetMicrosecondsPart(TimeSUB) == MS, "sub-seconds (%lu) = Sub2MicroSecs (%lu)", SUB, MS);
```

Comment:
This test is calling the same methods that the function it's testing is which is not ideal. Would be better if you started with some specific subsecond value, call the function, and then compare that with the known microsecond value.

Reply:
Ok, but what is the factor for calculating sub into microseconds? For example, if I have SUB = 400.